cupy.kron can accept numeric arguments, replicating numpy.kron behavior

sametz opened the pull request:
Re: #7596 `cupy.kron` now accepts numeric types (integers, floats) as arguments, which replicates `numpy.kron` behavior.

This is a WIP because my tests are written as pure pytest, and not using unittest as the developer's guide requires. I can see there is a lot of magic in CuPy's testing suite. I wanted to get feedback first to make sure I'm not barking up the wrong tree.

takagi commented on cupy/linalg/_product.py:
```diff
@@ -391,6 +391,14 @@ def kron(a, b):
     .. seealso:: :func:`numpy.kron`
 
     """
+    import numbers
```
Please put `import` at the beginning part of the file.

takagi commented:
> and not using unittest as the developer's guide requires

Whoops, the developer's guide is outdated and we moved from `unittest` to `pytest` for our testing module. Test codes in pytest is okay.

takagi commented:
Would you move your test in `TestProduct` class in `test_product.py`? It already has some `kron` tests. Also, you may use `numpy_cupy_allclose()` to check if NumPy and CuPy return equivalent results. You can find how to use it in the `kron` test in `TestProduct`. Thanks!

sametz commented:
- I moved the import where it belongs. 
- I simplified the test and moved it to TestProduct. 
- It is using `numpy_cupy_allclose()`. 
- Instead of using `pytest` parametrization, I refactored the test to use unittest.TestCase.subTest().

takagi commented on tests/cupy_tests/linalg_tests/test_product.py:
```diff
@@ -377,6 +377,16 @@ def test_zerodim_kron(self, xp, dtype):
         return xp.kron(a, b)
 
+    @testing.numpy_cupy_allclose()
+    def test_kron_accepts_numbers_as_arguments(self, xp):
```
You can use `@pytest.mark.parametrize` at the same time with `@testing.numpy_cupy_allclose` here. Did you try it and have some problems?

takagi commented on tests/cupy_tests/linalg_tests/test_product.py:
```diff
@@ -377,6 +377,16 @@ def test_zerodim_kron(self, xp, dtype):
         return xp.kron(a, b)
 
+    @testing.numpy_cupy_allclose()
+    def test_kron_accepts_numbers_as_arguments(self, xp):
```
We prefer using `@pytest.mark.parametrize` to `subTest` you suggested. Can I ask you to have it here? Thanks!

sametz replied:
I should be able to do this today. (Initially I tried an inner pytest decorator, but it had references to `xp.kron` within it, which failed.) With this refactoring, should I make `@pytest.mark.parametrize` the outer decorator? This seems consistent with other tests.

takagi replied:
Outer one is fine.

sametz replied:
I gave a try, and whether the pytest decorator is outer or inner, it doesn't seem to be passing its arguments to the test function. For example, as an outer decorator:
```python
    @pytest.mark.parametrize(
        "a, b", [
            (2, 3.0),
            (2, [[0, -1j / 2], [1j / 2, 0]]),
            ([[0, -1j / 2], [1j / 2, 0]], 2)
        ]
    )
    @testing.numpy_cupy_allclose()
    def test_kron_accepts_numbers_as_arguments(self, a, b, xp):
        args = [xp.array(arg) if type(arg) == list else arg for arg in [a, b]]
        print('converted args are ', args)   # print debug to be deleted before committing
        return xp.kron(*args)
```
results in the error
```
TypeError: TestProduct.test_kron_accepts_numbers_as_arguments() missing 2 required positional arguments: 'a' and 'b'
```

Full error below:
```
tests\cupy_tests\linalg_tests\test_product.py:379 (TestProduct.test_kron_accepts_numbers_as_arguments)
..\..\..\cupy\testing\_loops.py:273: in test_func
    _call_func_numpy_cupy(
..\..\..\cupy\testing\_loops.py:94: in _call_func_numpy_cupy
    cupy_result, cupy_error = _call_func_cupy(
..\..\..\cupy\testing\_loops.py:70: in _call_func_cupy
    result, error = _call_func(impl, args, kw)
..\..\..\cupy\testing\_loops.py:51: in _call_func
    raise e
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

impl = <function TestProduct.test_kron_accepts_numbers_as_arguments at 0x00000284BDFCD4E0>
args = (<cupy_tests.linalg_tests.test_product.TestProduct testMethod=test_kron_accepts_numbers_as_arguments>,)
kw = {'xp': <module 'cupy' from 'D:\\Documents\\GitHub\\cupy\\cupy\\__init__.py'>}

    def _call_func(impl, args, kw):
        # Note that `_pytest.outcomes.Skipped` is derived from BaseException.
        exceptions = Exception,
        if _is_pytest_available:
            exceptions += _pytest.outcomes.Skipped,
    
        try:
>           result = impl(*args, **kw)
E           TypeError: TestProduct.test_kron_accepts_numbers_as_arguments() missing 2 required positional arguments: 'a' and 'b'

..\..\..\cupy\testing\_loops.py:45: TypeError
```

takagi replied:
Ah, `@pytest.mark.parametrize` does not live in a `unittest.TestCase`'s subclass. Would you also make `TestProduct` class not deriving from `unittest.TestCase`? pytest needs no base testcase class.

sametz replied:
I moved the test into its own `TestKron` class, allowing `@pytest.mark_parametrize` to be used. Note the double commit because I almost forgot to remove my print debug statement. 🤦🏽‍♂️

takagi reviewed (approved):
LGTM!

takagi commented:
/test mini

takagi commented:
I added a slight change to your branch. Will merge after CI finishes.

takagi commented:
Thanks, @sametz!

sametz commented:
You're welcome! This is officially my first ever contribution to open-source software.

I see now what you meant by "Would you also make TestProduct class not deriving from unittest.TestCase?". 👍
